torchtune/_cli/run.py

```diff
@@ -0,0 +1,159 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+import argparse
+import runpy
+import sys
+import textwrap
+
+from pathlib import Path
+from typing import Optional
+
+import torchtune
+
+from torch.distributed.run import get_args_parser as get_torchrun_args_parser, run
+from torchtune import Config, get_all_recipes, Recipe
+from torchtune._cli.subcommand import Subcommand
+
+ROOT = Path(torchtune.__file__).parent.parent
+
+
+class Run(Subcommand):
+    """Holds all the logic for the `tune run` subcommand."""
+
+    def __init__(self, subparsers):
+        super().__init__()
+        self._parser = subparsers.add_parser(
+            "run",
+            prog="tune run",
+            help="Run a recipe. For distributed recipes, this supports all torchrun arguments.",
+            description="Run a recipe. For distributed recipes, this supports all torchrun arguments.",
+            usage="tune run [TORCHRUN-OPTIONS] <recipe> --config <config> [RECIPE-OPTIONS]",
+            epilog=textwrap.dedent(
+                """\
+                examples:
+
+                    # Run a finetuning recipe on a single device w/ default values
+                    $ tune run lora_finetune_single_device --config llama2/7B_lora_single_device
+
+                    # Run a finetuning recipe in a distributed fashion using torchrun w/ default values
+                    $ tune run --nproc_per_node 4 full_finetune_distributed --config llama2/7B_full_finetune_distributed
+
+                    # Override a parameter in the config file and specify a number of GPUs for torchrun
+                    $ tune run lora_finetune_single_device \
+                        --config llama2/7B_lora_single_device \
+                        model.lora_rank=16 \
+
+                Remember, you can use `tune cp` to copy a default recipe/config to your local dir and modify the values.
+                """
+            ),
+            formatter_class=argparse.RawTextHelpFormatter,
+        )
+        self._add_arguments()
+        self._parser.set_defaults(func=self._run_cmd)
+
+    def _add_arguments(self) -> None:
+        """Add arguments to the parser."""
+        torchrun_argparser = get_torchrun_args_parser()
+        # Add torchrun arguments to the parser
+        for action in torchrun_argparser._actions:
+            if action.dest == "training_script":
+                action.dest = "recipe"
+                action.help = """
+                    Name or path to recipe to be launched followed by args.
+                    For a list of all possible recipes, run `tune ls`."""
+            elif action.dest == "training_script_args":
+                action.dest = "recipe_args"
+                action.help = "Args to be passed to the recipe."
+            elif action.dest == "help":
+                continue
```

Comment:
I think I understand this? But still, might be nice to give a bit more detail in the docstring here